text/010-wasm-pack-watch-and-serve.md

```diff
@@ -0,0 +1,230 @@
+port the local server should bind to.
+
+```
+USAGE:
+    wasm-pack serve [FLAGS] [OPTIONS] [path] [-- <extra_options>...]
+
+FLAGS:
+        --dev              Create a development build. Enable debug info, and disable optimizations.
+        --no-typescript    By default a *.d.ts file is generated for the generated JS file, but this flag will disable
+                           generating this TypeScript file.
+    -h, --help             Prints help information
+        --profiling        Create a profiling build. Enable optimizations and debug info.
+        --release          Create a release build. Enable optimizations and disable debug info.
+    -V, --version          Prints version information
+        --no-watch         Do not watch the crate for changes to automaticaly rebuild.
+
+OPTIONS:
+    -m, --mode <mode>          Sets steps to be run. [possible values: no-install, normal, force] [default: normal]
+    -d, --out-dir <out_dir>    Sets the output directory with a relative path. [default: pkg]
+    -s, --scope <scope>        The npm scope to use in package.json, if any.
+    -t, --target <target>      Sets the target environment. [possible values: browser, nodejs, no-modules] [default:
+                               browser]
+    -p, --port <port>          Bind to this port number with the local development HTTP server. [default: 8000]
+
+ARGS:
+    <path>                The path to the Rust crate.
+    <extra_options>...    List of extra options to pass to `cargo build`
+```
+
+### Implementation
+
+Rather than integrating an HTTP server into the `wasm-pack` binary itself, we
+should leverage its ability to download and run other binaries. Exactly *which*
+local HTTP server binary is left as an unresolved question.
+
+The subcommand will start the local HTTP server and then execute the `watch`
+subcommand (unless `--no-watch` is supplied).
+
+# Drawbacks, Rationale, and Alternatives
+
+The primary drawbacks are authoring and maintaining two more subcommands to
+implement functionality that is already available in external, third-party
+tools. We try to mitigate this downside by using existing local HTTP server
+binaries rather than building an HTTP server directly into `wasm-pack` itself.
+
+## Alternative: Build the HTTP Server into `wasm-pack`
+
+An alternative design would be to build the HTTP server into `wasm-pack` itself,
+using one of the existing crates for building Web servers like rocket, actix,
+hyper, tide, etc:
+
+* **Pros:**
+  * The `wasm-pack serve` command is ready to roll immediately after `wasm-pack`
+    is installed, and doesn't need to hit the network the first time you run
+    it. This situation only arises when folks who already downloaded `wasm-pack`
+    do their first build/serve offline, disconnected from the internet. This
+    seems like a niche enough situation that we can take the trade off.
+* **Cons:**
+  * Building the HTTP server into the `wasm-pack` binary is less robust: by
```

Comment:
FWIW this doesn't seem like too much of a con to me, if we were dealing with C/C++ I'd definitely agree but with Rust error recovery is pretty simple in-process. 

That being said I still agree the next con outweighs the pro

Reply:
I wasn't thinking just memory, but also other resources like sockets, but yeah much less of a deal in Rust.